### PR TITLE
feat: QC key aggregate gadget

### PR DESCRIPTION
**Pull request by mrain**
closes: #71 

- Added: QC key aggregation gadgets for short Weierstrass curve points
- Unified key aggregation gadget interfaces

### CODE REVIEW

File: src/circuit/qc_keyagg_sw.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
+//! Circuit implementation of stake key aggregation for quorum certificates verification.
```
Comment (chancharles92) [inline src/circuit/qc_keyagg_sw.rs:1]: Update header comment accordingly

File: src/circuit/qc_keyagg_sw.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+};
+
+#[derive(Debug, Clone)]
+/// Stake public key variable
+/// Wrap EmulatedSWPointVariable because we need to simulate curve operations over the same curve's scalar field
+pub struct VerKeySWVar<E: PrimeField>(pub EmulatedSWPointVariable<E>);
+
+/// Plonk circuit gadget for stake key aggregation for quorum certificates.
+/// Assuming that the underlying key composes of a short Weierstrass curve point.
+pub trait QCKeyAggregateSWGadget<F>
```
Comment (chancharles92) [inline src/circuit/qc_keyagg_sw.rs:25]: Can we have a unified trait for both SW points and TE points?

The implementations also seem pretty similar, is it possible to have a generic implementation?

File: src/circuit/qc_keyagg_sw.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                );
+        let agg_vk_point: SWPoint<E> = agg_vk_point.into_affine().into();
+        let vk_points: Vec<SWPoint<E>> = vk_points.iter().map(|p| p.into_affine().into()).collect();
+        let stake_amts: Vec<F> = (0..5).map(|i| F::from((i + 1) as u32)).collect();
+        let threshold = F::from(6u8);
+        let digest = compute_stake_table_hash::<F, E>(&stake_amts[..], &vk_points[..]);
+
+        let mut circuit = PlonkCircuit::<F>::new_ultra_plonk(20);
+        // public input
+        // TODO: make variables public?
```
Comment (chancharles92) [inline src/circuit/qc_keyagg_sw.rs:195]: Can we fix this TODO?

File: src/circuit/qc_keyagg_sw.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            .is_err());
+        assert!(circuit
+            .check_threshold(&stake_amt_vars[..], &bitvec_vars[1..], threshold_var)
+            .is_err());
+
+        Ok(())
+    }
+
+    // TODO(binyi): make the API public
+    fn compute_stake_table_hash<F: RescueParameter, E: EmulationConfig<F>>(
```
Comment (chancharles92) [inline src/circuit/qc_keyagg_sw.rs:251]: Move this out of the test code, and make it a public APIs for users.

File: src/circuit/qc_keyagg.rs [hunk 5]
```diff
@@ -1,26 +1,71 @@
...
+        EmulationConfig, SerializableEmulatedStruct,
     },
     BoolVar, Circuit, PlonkCircuit, Variable,
 };
 
-#[derive(Debug, Clone)]
-/// Stake public key variable
-/// Wrap EmulatedPointVariable because we need to simulate curve operations over the same curve's scalar field
-pub struct VerKeyVar<E: PrimeField>(pub EmulatedPointVariable<E>);
+pub fn compute_stake_table_hash<F: RescueParameter, T: SerializableEmulatedStruct<F>>(
```
Comment (chancharles92) [inline src/circuit/qc_keyagg.rs:21]: Add some comment for public APIs.

File: src/circuit/qc_keyagg.rs [hunk 6]
```diff
@@ -1,26 +1,71 @@
...
+        p0: &Self,
+        p1: &Self,
+    ) -> Result<(), CircuitError>
+    where
+        F: PrimeField,
+        E: EmulationConfig<F>;
+}
 
 /// Plonk circuit gadget for stake key aggregation for quorum certificates.
+/// Assuming that the underlying key composes of a short Weierstrass curve point.
```
Comment (chancharles92) [inline src/circuit/qc_keyagg.rs:68]: Do we want to have this assumption for a general trait?

**Review by chancharles92 — COMMENTED** [review]
Good job and LGTM, only 2 minor comments

File: src/circuit/qc_keyagg.rs (hunk 5 above)
Comment (mrain) [reply src/circuit/qc_keyagg.rs:24]: Done in [8697221](https://github.com/EspressoSystems/hotshot-primitives/pull/74/commits/869722169145e40b275b9d16e50284a3868e2a60)

File: src/circuit/qc_keyagg.rs (hunk 6 above)
Comment (mrain) [reply src/circuit/qc_keyagg.rs]: Removed in [8697221](https://github.com/EspressoSystems/hotshot-primitives/pull/74/commits/869722169145e40b275b9d16e50284a3868e2a60)
 No assumption now.

File: src/circuit/qc_keyagg.rs [hunk 7]
```diff
@@ -141,42 +192,169 @@ where
...
+        errors::CircuitError, gadgets::ecc::SWToTEConParam, Circuit, PlonkCircuit, Variable,
     };
 
     #[test]
-    fn test_vk_aggregate_circuit() -> Result<(), CircuitError> {
-        let d_ecc : Fq377 = MontFp!("122268283598675559488486339158635529096981886914877139579534153582033676785385790730042363341236035746924960903179");
-        test_vk_aggregate_circuit_helper::<Fq377, Fr254, Param377>(d_ecc)
+    fn test_vk_aggregate_sw_circuit() -> Result<(), CircuitError> {
+        let a_ecc = Fq377::zero();
+        test_vk_aggregate_sw_circuit_helper::<Fq377, Fr254, Param377>(a_ecc)?;
```
Comment (chancharles92) [inline src/circuit/qc_keyagg.rs:338]: Why are we using Fr254 rather than Fr377 here?

File: src/circuit/qc_keyagg.rs [hunk 8]
```diff
@@ -187,85 +365,180 @@ mod tests {
...
-            let y_vec = from_emulated_field(point.get_y());
-            input_vec.extend(x_vec);
-            input_vec.extend(y_vec);
-            input_vec.push(amt);
-        }
-        RescueCRHF::sponge_with_bit_padding(&input_vec[..], 1)[0]
+    #[test]
+    fn test_vk_aggregate_te_circuit() -> Result<(), CircuitError> {
+        let d_ecc : Fq377 = MontFp!("122268283598675559488486339158635529096981886914877139579534153582033676785385790730042363341236035746924960903179");
+        test_vk_aggregate_te_circuit_helper::<Fq377, Fr254, Param377>(d_ecc)
```
Comment (chancharles92) [inline src/circuit/qc_keyagg.rs:444]: Same here, should we replace Fr254 with Fr377?

File: src/circuit/qc_keyagg.rs (hunk 7 above)
Comment (mrain) [reply src/circuit/qc_keyagg.rs:338]: We only have emulation config for native field `Fr254`. IMO that's sufficient, nothing special for `Fr377`.

File: src/circuit/qc_keyagg.rs (hunk 8 above)
Comment (mrain) [reply src/circuit/qc_keyagg.rs:444]: Same as above

**Review by chancharles92 — APPROVED** [review]

**Pull request merged**